base/threadingconstructs.jl

```diff
@@ -110,15 +112,48 @@ A barrier is placed at the end of the loop which waits for all tasks to finish
 execution.
 
 The `schedule` argument can be used to request a particular scheduling policy.
-The only currently supported value is `:static`, which creates one task per thread
-and divides the iterations equally among them. Specifying `:static` is an error
-if used from inside another `@threads` loop or from a thread other than 1.
+Options are:
+- `:static` is the default schedule which creates one task per thread and divides the
+            iterations equally among them, assigning each task specifically to each thread.
+            Specifying `:static` is an error if used from inside another `@threads` loop
+            or from a thread other than 1.
+- `:dynamic` is like `:static` except the tasks are assigned to threads dynamically,
+            allowing more flexible scheduling if other tasks are active on other threads.
+            Specifying `:dynamic` is allowed from inside another `@threads` loop and from
+            threads other than 1.
```

Comment:
I suggest avoiding specifying the behavior as much as possible, even though it may be a losing battle against Hyrum's law.

Before mentioning schedules, I think it'd be a good idea to explicitly say what is OK and what is not. Maybe something like:

> Except `:static` scheduling, how the iterations are assigned to tasks, and how the tasks are assigned to the worker threads are undefined can be different for each execution. The scheduling option should be considered as a hint. The loop body code (including any code transitively called from it) must not assume the task and worker thread in which they are executed. The loop body code for each iteration must be able to make forward progress and be free from data races independent of the state of other iterations. As such, synchronizations across iterations may invoke deadlock.
>
> For example, the above conditions imply that:
> * The lock taken in an iteration must be released within the same iteration.
> * Avoid communication between iterations using, e.g., `Channel`s.
> * Write only to locations not shared across iterations (unless lock or atomic operation is used).
>
> Furthermore, even though lock and atomic can be useful sometimes, it is often better to avoid them for performance.

(We can maybe borrow some ideas from the C++ executor definition to make the wording more precise.)

After this, I think we can simply mention

```suggestion
- `:dynamic` tries to schedule iterations dynamically to available worker threads,
            assuming that the workload for each iteration is uniform.
```

...which makes me wonder if `:uniform` is a better wording? Users may expect some kind of load-balancing for `:dynamic` scheduling.

Reply:
Makes sense. I've pushed your suggestions. 

I think `:dynamic` makes a bit more sense as it's a description of the scheduling policy, not the underlying workload, which `:uniform` refers to?

Reply:
My thought was more like, since possible additional scheduling will likely be all "dynamic" in some sense, it'd be better to emphasize other aspects. So yeah, it does focus on the underlying workload.

For example, it's reasonable to add `:loadbalancing` that splits the iteration space to `nthreads() * 8` chunks (say). But both `:dynamic` and `:loadbalancing` are "dynamic."

Reply:
That makes sense I think. If `:loadbalaning` was added the schedules could be summarized in the docstring as
```
- static

Dynamic options:
- uniform
- loadbalanced
```

Though, I think an issue is that in code `@threads :uniform for` doesn't alone speak to dynamic behavior, but perhaps that's ok?

Reply:
If this option becomes the "go to option" for people who want to do threading (which it should probably be if I understand correctly?), calling it uniform is a bit confusing (unless it's the default and can be omitted)

Reply:
IMHO "go to option" is FLoops.jl but it's really hard to tell people to not think "it's in `Base` so it must be good" (but yeah, of course, *I* would say this :laughing:). But FLoops.jl is not ready for `Base` and it's kind of stupid to not harvest *really* low-hanging fruit like this. That's why I thought that a patch like this is good.

Anyway, yes, I agree `:uniform` doesn't shout "USE ME!". Also, it may be reasonable to argue that even `:uniform` is specifying too much. Maybe we should just have `:dynamic`. It let us, in the future, internally use something similar to `:loadbalancing` if the compiler sees some branches and something similar to `:uniform` otherwise. Or maybe we can do some clever run-time trick.

Then, maybe we should simply write:

> New code using `@threads`, especially those that can be called from library API, should try to use `:dynamic`. It schedules iterations dynamically to available worker threads while minimizing overhead.

Reply:
What's the rationale for not making it the default? Is it safety considerations? If yes, might be good to make it explicit, eg "The default is :static, to ensure safety, but code with independent iterations should use :dynamic which is usually more performant."

Reply:
`:static` is not the default. It also does not ensure safety. The purpose of `:static` is to keep the pre-1.5 behavior in order to help the old programs that assumed the way `@threads` scheduled iterations in the previous versions.